peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+					"peer %v -- disconnecting", p)
+				break out
+			}
+			p.verAckReceived = true
+
+		case *wire.MsgGetAddr:
+			for key, listener := range p.getAddrMsgListeners {
+				log.Tracef("Running %s listener %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgAddr:
+			for key, listener := range p.addrMsgListeners {
+				log.Tracef("Running %s listener %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgPing:
+			p.handlePingMsg(msg)
+
+		case *wire.MsgPong:
+			p.handlePongMsg(msg)
+
+		case *wire.MsgAlert:
+			// Intentionally ignore alert messages.
+			//
+			// The reference client currently bans peers that send
+			// alerts not signed with its key.  We could verify
+			// against their key, but since the reference client
+			// is currently unwilling to support other
+			// implementions' alert messages, we will not relay
+			// theirs.
+
+		case *wire.MsgMemPool:
+			for key, listener := range p.memPoolMsgListeners {
+				log.Tracef("Running %s listener %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgTx:
+			for key, listener := range p.txMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgBlock:
+			for key, listener := range p.blockMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg, buf)
+			}
+
+		case *wire.MsgInv:
+			for key, listener := range p.invMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgHeaders:
+			for key, listener := range p.headersMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgNotFound:
+			// TODO(davec): Ignore this for now, but ultimately
+			// it should probably be used to detect when something
+			// we requested needs to be re-requested from another
+			// peer.
+
+		case *wire.MsgGetData:
+			for key, listener := range p.getDataMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgGetBlocks:
+			for key, listener := range p.getBlocksMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgGetHeaders:
+			for key, listener := range p.getHeadersMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgFilterAdd:
+			for key, listener := range p.filterAddMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgFilterClear:
+			for key, listener := range p.filterClearMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgFilterLoad:
+			for key, listener := range p.filterLoadMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgReject:
+			// Nothing to do currently.  Logging of the rejected
```

Comment:
Same here.  We should expose a listener for all supported messages even if btcd doesn't currently directly use them since this peer is meant for more than just btcd.